### ISSUE TITLE
setting for bowtie2 mapping

### ISSUE THREAD

**Issue by aistBMRG** [RasmussenLab/vamb#83]
Hi,

Thanks for developing vamb; excited to try it.

Now, it seems that the mapping should be performed carefully, more so than for other binners I guess.

It seems that you often used minimap2. Would bowtie2 also be appropriate for short reads? Do you have any experience with that, and which settings (e.g., --very-sensitive) would be suitable for the alignments?

Thanks.

Dieter

**Comment by simonrasmu** [comment]
Hi Dieter,

We haven't tried bowtie2, but feel free to try it out and let us know how it works. If your dataset is not that big I would try with both minimap2 and bowtie2 (or different parameters of bowtie2) and compare the outcome.

Best,

Simon

**Comment by aistBMRG** [comment]
Thanks. I have a very big dataset (>1200 metagenomes) so will be  sticking with minimap2 as per your workflow.

Looking at the snakemake workflow, I had a small question. For the quantification using the jgi method, it seems that the final abundance matrix is build by combining (pasting) the results across different samples. Now, I was wondering as to why columns 1 to 3 are kept for the first sample, the third column is the "totalAvgDepth" and is the same as the depth of the first sample since samples/bam files are quantified separately. Just wanted to make sure that is as intended. Below is an example as generated (jgi.abundance.dat).

contigName | contigLen | totalAvgDepth | H53G2DSXY_PG3460_759A5454_H1_L001.sort.bam | H53G2DSXY_PG3460_760A5656_H1_L001.sort.bam | H53G2DSXY_PG3460_761A5757_H1_L001.sort.bam
-- | -- | -- | -- | -- | --
S1Ck141_15295 | 2026 | 3.10608 | 3.10608 | 10.9174 | 0
S1Ck141_36225 | 2049 | 1.41654 | 1.41654 | 9.06372 | 0
S1Ck141_1 | 2397 | 257.682 | 257.682 | 5.37072 | 1.4664
S1Ck141_33005 | 2850 | 7.65741 | 7.65741 | 0 | 5.54815
S1Ck141_35421 | 2190 | 2.79608 | 2.79608 | 0 | 0.642647
S1Ck141_22541 | 2396 | 2.09751 | 2.09751 | 0 | 1.33482
S1Ck141_2417 | 3401 | 8.51584 | 8.51615 | 0.0901261 | 0
S1Ck141_27371 | 4169 | 3.91565 | 3.91565 | 8.05474 | 0.362528
S1Ck141_5635 | 4936 | 3.01149 | 3.01149 | 0.124948 | 3.68094

Also, just wanted to ensure that the renamed samples in the samples2data file, that is the first sample is S1 etc.

Thanks.

Dieter

**Comment by simonrasmu** [comment]
Hi Dieter,

Yes - you are correct. The first three columns are not used by VAMB, but are imposed by the jgi-format. Therefore the "totalAvgDepth" is not correct (this is for the first sample only), but it does not affect VAMB processing. The main reason for doing it in this way is that it is _much_ faster (fully parallelizable) compared to running all in one go - even running in one go might not always finished and can crash due to memory etc.

Yes, the first sample is S1, but pinging @jakobnissen to confirm

Best,

Simon

**Comment by aistBMRG** [comment]
Hi Simon,

Thanks for the clarification. The main issue I am facing now will be the time needed for mapping, mainly due to the need to read the index (~0.5 Tb unipart index) into RAM for each sample. That takes most of the time and not sure how we can keep the index in RAM without the need to reimport for each sample .... did you run into this issue?

Thanks for developing VAMB, seems a great tool.

Dieter

**Comment by simonrasmu** [comment]
Hi Dieter,

Ah ok, that is a lot of data. The amount of memory that your mapping uses can be controlled using the option `index_size` in the workflow - this corresponds to `-I` when using `minimap2 index` command. We found that `-I` of 12G corresponds to roughly 35Gb RAM usage, you can modify it as you like. However, this means that minimap2 will automatically map to subsets of the contigs in turn and this has some impact for when it is mapping. This is all taken care of in the snakemake (see the rule minimap) just writing it so you know if you use it outside of the workflow.

If it is still too much data make sure that you have:
- Removed contigs under 2kbp. We found that including smaller contigs just makes it _much_ slower and we don't get any more high quality MAGs out of it. This is mainly because it is difficult to estimate TNF from small contigs.
- Split it into two sets. Basically run the VAMB pipeline on only 600 samples each and then afterwards combine MAGs using dRep or Galah.

Best,

Simon

**Comment by aistBMRG** [comment]
Hi Simon,

Thanks for the response and taking time for clarifying. 

Actually, we have enough memory (1 Tb) to load the entire index into RAM (takes about 500 Gb); it is just that this loading step takes a long time and is rate-limiting as it is performed for each sample. Not sure if we can tell minimap2 to keep using the same index in RAM for different samples ...

Yes, I prepared a single index (with -I 900G) since I guess mapping to a multi-part index may lead to suboptimal results as you pointed out. Yes, I discarded small (<2000 bp) contigs, as part of the concatenation script that is provided.

At this point, I will just run the mapping on several of our machines in parallel, that will just take time ... I can then generate the jgi abundance matrix to run the machine learning part (I guess that may be ok since we have 500 Gb of RAM on a machine with GPU).

Thanks.

Dieter

**Comment by simonrasmu** [comment]
The issue with multi-part index is taken care of by the snakemake-workflow, so it should be possible to run it using much smaller indexes. So it should work automatically, my suggestion is to try it out on a few samples.

Best,

Simon

**Comment by aistBMRG** [comment]
Ok, thanks. I will check the minimap2 rule and then run with a smaller batch of samples.

Thanks again,

Dieter
